File: php/src/Cache.php
<?php

namespace LeastRecentlyUsed;

class Cache {
	private $capacity;
	private $cache = [];
	private $head = null;
	private $tail = null;

	/**
	 * Constructor for Cache
	 * @param int $capacity The maximum number of items the cache can hold
	 * @throws \InvalidArgumentException if capacity is not 1+
	 */
	public function __construct(int $capacity) {
		if ($capacity <= 0) throw new \InvalidArgumentException("Cache capacity must be at least 1");
		$this->capacity = $capacity;
	}

	/**
	 * Get an item from the cache
	 * @param string $key The key to look up
	 * @return mixed|null The value if found, null otherwise
	 */
	public function get(string $key): mixed {
		if (!isset($this->cache[$key])) return null;

		$node = $this->cache[$key];
		$this->moveToFront($node);
		return $node->value;
	}

	/**
	 * Put an item into the cache
	 * @param string $key The key to store the value under
	 * @param mixed $value The value to store
	 */
	public function put(string $key, mixed $value): void {
		if (isset($this->cache[$key])) {
			$node = $this->cache[$key];
			$node->value = $value;
			$this->moveToFront($node);
			return;
		}

		$node = new CacheItem($key, $value);
		$this->cache[$key] = $node;
		$this->addToFront($node);

		if (count($this->cache) > $this->capacity) $this->removeLast();
	}

	/**
	 * Move a node to the front of the list (most recently used)
	 * @param CacheItem $node The node to move
	 */
	private function moveToFront(CacheItem $node): void {
		if ($node === $this->head) return;

		// Remove node from its current position
		if ($node->prev) $node->prev->next = $node->next;
		if ($node->next) $node->next->prev = $node->prev;
		if ($node === $this->tail) $this->tail = $node->prev;

		$this->addToFront($node);
	}

	/**
	 * Add a node to the front of the list
	 * @param CacheItem $node The node to add
	 */
	private function addToFront(CacheItem $node): void {
		$node->next = $this->head;
		$node->prev = null;

		if ($this->head) $this->head->prev = $node;
		$this->head = $node;

		if ($this->tail === null) $this->tail = $node;
	}

	/**
	 * Remove the least recently used item from the cache
	 */
	private function removeLast(): void {
		if ($this->tail === null) return;

		$key = $this->tail->key;
		unset($this->cache[$key]);

		if ($this->head === $this->tail) {
			$this->head = null;
			$this->tail = null;
			return;
		}

		$this->tail = $this->tail->prev;
		$this->tail->next = null;
	}
}

File: go/lrucache/cache.go
package lrucache

import "errors"

// ErrInvalidCapacity is returned when trying to create a cache with capacity <= 0
var ErrInvalidCapacity = errors.New("capacity must be positive")

// Cache implements an LRU (Least Recently Used) cache
type Cache struct {
	capacity int
	cache    map[string]*CacheItem // key -> CacheItem mapping
	head     *CacheItem           // Most recently used
	tail     *CacheItem           // Least recently used
}

// Capacity returns the maximum number of items the cache can hold
func (c *Cache) Capacity() int {
	return c.capacity
}

// New creates a new LRU Cache with the specified capacity
func NewCache(capacity int) (*Cache, error) {
	if capacity <= 0 {
		return nil, ErrInvalidCapacity
	}

	c := &Cache{
		capacity: capacity,
		cache:    make(map[string]*CacheItem),
	}

	// Initialize dummy head and tail nodes - these are used to simplify list manipulation
	c.head = NewCacheItem("", nil)
	c.tail = NewCacheItem("", nil)
	c.head.next = c.tail
	c.tail.prev = c.head

	return c, nil
}

// removeNode removes a node from its current position in the list
func (c *Cache) removeNode(item *CacheItem) {
	item.prev.next = item.next
	item.next.prev = item.prev
}

// addToHead adds a node to the head of the list (most recently used)
func (c *Cache) addToHead(item *CacheItem) {
	item.next = c.head.next
	item.prev = c.head
	c.head.next.prev = item
	c.head.next = item
}

// moveToHead moves an existing node to the head of the list
func (c *Cache) moveToHead(item *CacheItem) {
	c.removeNode(item)
	c.addToHead(item)
}

// Get retrieves an item from the cache by key
func (c *Cache) Get(key string) (interface{}, bool) {
	if item, exists := c.cache[key]; exists {
		c.moveToHead(item)
		return item.value, true
	}
	return nil, false
}

// Put adds or updates an item in the cache
func (c *Cache) Put(key string, value interface{}) {
	if item, exists := c.cache[key]; exists {
		// Update existing node
		item.value = value
		c.moveToHead(item)
		return
	}

	// Create new node
	item := NewCacheItem(key, value)
	c.cache[key] = item
	c.addToHead(item)

	// Remove least recently used if at capacity
	if len(c.cache) > c.capacity {
		lruItem := c.tail.prev
		c.removeNode(lruItem)
		delete(c.cache, lruItem.key)
	}
}
